sys/net/transport_layer/ng_udp/ng_udp.c

```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright (C) 2015 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     net_ng_udp
+ * @{
+ *
+ * @file
+ * @brief       UDP implementation
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @}
+ */
+
+#include <stdint.h>
+#include <errno.h>
+
+#include "kernel.h"
+#include "byteorder.h"
+#include "msg.h"
+#include "thread.h"
+#include "utlist.h"
+#include "net/ng_udp.h"
+#include "net/ng_netbase.h"
+#include "net/ng_inet_csum.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+/**
+ * @brief   Save the UDP's thread PID for later reference
+ */
+static kernel_pid_t _pid = KERNEL_PID_UNDEF;
+
+/**
+ * @brief   Allocate memory for the UDP thread's stack
+ */
+static char _stack[NG_UDP_STACK_SIZE];
+
+/**
+ * @brief   Calculate the UDP checksum dependent on the network protocol
+ *
+ * @note    If the checksum turns out to be 0x0000, the function returns 0xffff
+ *          as specified in RFC768
+ *
+ * @param[in] pkt           pointer to the packet in the packet buffer
+ * @param[in] pseudo_hdr    pointer to the network layer header
+ * @param[in] payload       pointer to the payload
+ *
+ * @return                  the checksum of the pkt in host byte order
+ * @return                  0 on error
+ */
+static uint16_t _calc_csum(ng_pktsnip_t *hdr, ng_pktsnip_t *pseudo_hdr,
+                           ng_pktsnip_t *payload)
+{
+    uint16_t csum;
+    /* process the payload */
+    csum = ng_inet_csum(0, (uint8_t *)(payload->data), payload->size);
+    /* process applicable UDP header bytes */
+    csum += ng_inet_csum(csum, (uint8_t *)hdr->data, 6);
+
+    switch (pseudo_hdr->type) {
+#ifdef MODULE_NG_IPV6
+        case NG_NETTYPE_IPV6:
+            csum += ng_inet_csum(csum, ((uint8_t *)pseudo_hdr) + 8, 32);
+            csum += byteorder_htons((uint16_t)(hdr->size + payload->size)) +
+                    ((ng_ipv6_hdr_t *)pseudo_hdr->data)->nh;
```

Comment:
ah, I had the impression that you implemented this already, just couldn't find it... But that function is not yet merged, right?